src/data/gaussian_int.lean

```diff
@@ -0,0 +1,170 @@
+
+lemma to_complex_def₂ (x : ℤ[i]) : (x : ℂ) = ⟨x.re, x.im⟩ :=
+by apply complex.ext; simp [to_complex_def]
+
+instance to_complex.is_ring_hom : is_ring_hom to_complex:=
+by refine_struct {..}; intros; apply complex.ext; simp [to_complex]
+
+instance : is_ring_hom (coe : ℤ[i] → ℂ) := to_complex.is_ring_hom
+
+@[simp] lemma int_cast_re (x : ℤ[i]) : ((x.re : ℤ) : ℝ) = (x : ℂ).re := by simp [to_complex_def]
+@[simp] lemma int_cast_im (x : ℤ[i]) : ((x.im : ℤ) : ℝ) = (x : ℂ).im := by simp [to_complex_def]
+@[simp] lemma to_complex_re' (x y : ℤ) : ((⟨x, y⟩ : ℤ[i]) : ℂ).re = x := by simp [to_complex_def]
+@[simp] lemma to_complex_im' (x y : ℤ) : ((⟨x, y⟩ : ℤ[i]) : ℂ).im = y := by simp [to_complex_def]
+@[simp] lemma to_complex_add (x y : ℤ[i]) : ((x + y : ℤ[i]) : ℂ) = x + y := is_ring_hom.map_add coe
+@[simp] lemma to_complex_mul (x y : ℤ[i]) : ((x * y : ℤ[i]) : ℂ) = x * y := is_ring_hom.map_mul coe
+@[simp] lemma to_complex_one : ((1 : ℤ[i]) : ℂ) = 1 := is_ring_hom.map_one coe
+@[simp] lemma to_complex_zero : ((0 : ℤ[i]) : ℂ) = 0 := is_ring_hom.map_zero coe
+@[simp] lemma to_complex_neg (x : ℤ[i]) : ((-x : ℤ[i]) : ℂ) = -x := is_ring_hom.map_neg coe
+@[simp] lemma to_complex_sub (x y : ℤ[i]) : ((x - y : ℤ[i]) : ℂ) = x - y := is_ring_hom.map_sub coe
+
+@[simp] lemma to_complex_inj {x y : ℤ[i]} : (x : ℂ) = y ↔ x = y :=
+by cases x; cases y; simp [to_complex_def₂]
+
+@[simp] lemma to_complex_eq_zero {x : ℤ[i]} : (x : ℂ) = 0 ↔ x = 0 :=
+by rw [← to_complex_zero, to_complex_inj]
+
+@[simp] lemma nat_cast_real_norm (x : ℤ[i]) : (x.norm : ℝ) = (x : ℂ).norm_sq :=
+by rw [norm, nat.cast_add, ← int.cast_coe_nat, ← int.cast_coe_nat,
+  int.nat_abs_mul_self, int.nat_abs_mul_self, complex.norm_sq]; simp
+
+@[simp] lemma nat_cast_complex_norm (x : ℤ[i]) : (x.norm : ℂ) = (x : ℂ).norm_sq :=
+by rw [← of_real_nat_cast, nat_cast_real_norm]
+
+@[simp] lemma norm_mul (x y : ℤ[i]) : norm (x * y) = norm x * norm y :=
+(@nat.cast_inj ℂ _ _ _ _ _).1 $ by simp
+
+@[simp] lemma norm_eq_zero {x : ℤ[i]} : norm x = 0 ↔ x = 0 :=
+by rw [← @nat.cast_inj ℝ _ _ _]; simp
+
+lemma norm_pos {x : ℤ[i]} : 0 < norm x ↔ x ≠ 0 :=
+by rw [nat.pos_iff_ne_zero', ne.def, norm_eq_zero]
+
+lemma norm_nat_cast (x : ℕ) : norm x = x ^ 2 :=
+by simp [norm, nat.pow_two]
+
+@[simp] lemma norm_one : norm 1 = 1 := rfl
+@[simp] lemma norm_zero : norm 0 = 0 := rfl
+
+lemma norm_eq_one_iff {x : ℤ[i]} : norm x = 1 ↔ is_unit x :=
+⟨λ h, is_unit_iff_dvd_one.2 begin
+  rw [norm, nat.add_eq_one_iff, ← int.coe_nat_inj', ← int.coe_nat_inj',
+    ← int.coe_nat_inj', ← int.coe_nat_inj', int.nat_abs_mul_self,
+    int.nat_abs_mul_self] at h,
+  cases h,
+  { exact ⟨⟨0, -x.im⟩, by simp [zsqrtd.ext, *] at *⟩ },
+  { exact ⟨⟨x.re, 0⟩, by simp [zsqrtd.ext, *] at *⟩ }
+end,
+λ h, let ⟨y, hy⟩ := is_unit_iff_dvd_one.1 h in begin
+  have := congr_arg norm hy,
+  rw [norm_one, norm_mul, eq_comm, nat.mul_eq_one_iff] at this,
+  exact this.1
+end⟩
+
+protected def div (x y : ℤ[i]) : ℤ[i] :=
```

Comment:
Is this a standard definition of division for Gaussian integers?

Reply:
I think so. The idea is that you have be within a radius of one of the complex division in order for the division to be euclidean, so rounding to the nearest integer seems like the canonical choice.